ironfish/src/syncer.ts

```diff
@@ -99,46 +111,171 @@ export class Syncer {
+        // the state to measuring. This gives the syncer time to finish any
+        // in-flight requests
+        this.state = 'measuring'
+        if (this.loader) {
+          this.lastLoaderIdentity = this.loader.getIdentity()
+          await this.wait()
+        }
+        break
+      }
     }
 
-    this.eventLoopTimeout = setTimeout(() => this.eventLoop(), SYNCER_TICK_MS)
+    this.eventLoopTimeout = setTimeout(() => void this.eventLoop(), SYNCER_TICK_MS)
   }
 
-  findPeer(): void {
+  /**
+   * Chooses a peer to sync from based on measuring the connection of a random
+   * sampling of connected peers and begins syncing
+   */
+  async findPeer(currentPeerIdentity: Identity | null): Promise<void> {
     const head = this.chain.head
-
     if (!head) {
       return
     }
 
+    this.logger.debug('Syncer is beginning peer candidate measurements')
+    const measurementStart = BenchUtils.start()
+
     // Find all allowed peers that have more work than we have
     const peers = this.peerNetwork.peerManager
       .getConnectedPeers()
       .filter((peer) => peer.features?.syncing && peer.work && peer.work > head.work)
 
-    // Get a random peer with higher work. We do this to encourage
-    // peer diversity so the highest work peer isn't overwhelmed
-    // as well as helping to make sure we don't get stuck with unstable peers
-    if (peers.length > 0) {
-      const peer = ArrayUtils.sampleOrThrow(peers)
+    if (peers.length === 0) {
+      return
+    }
+
+    // If there is only one valid peer to sync from, there is no point in
+    // measuring the connection so begin syncing immediately
+    if (peers.length === 1) {
+      this.startSync(peers[0])
+      return
+    }
+
+    const syncCandidates = ArrayUtils.shuffle(peers)
+
+    // If we have been syncing from a peer, we want to include this peer in the
+    // measurement. This will allow us to maintain a connection to a strong peer
+    // if we have found one.
+    if (
+      currentPeerIdentity &&
+      !syncCandidates.find((p) => p.getIdentity() === currentPeerIdentity)
+    ) {
+      const currentPeer = this.peerNetwork.peerManager.getPeer(currentPeerIdentity)
+      if (currentPeer && currentPeer.state.type === 'CONNECTED') {
+        syncCandidates.push(currentPeer)
+      }
+    }
+
+    const peerRtt = new Map<Identity, number>()
+
+    // Measure how long it takes to fetch the genesis block header from each
+    // peer as an estimate of connection quality
+    let candidateCheckCount = 0
+    for (const peer of syncCandidates) {
+      if (this.state === 'stopped' || this.state === 'stopping') {
+        return
+      }
+
+      // We only want to successfully measure so many candidates per measurement
+      // phase
+      if (candidateCheckCount > 8) {
+        break
+      }
+
+      const peerIdentity = peer.getIdentity()
+      if (peerIdentity == null) {
+        continue
+      }
```

Comment:
What do you think about filtering these peers out before the loop starts? That way you have a better selection anyways

Reply:
Hm, good call-out, technically, we _are_ filtering these out, I just didn't think about this enough. A peer, if its state is connected, has a non-nullable identity field (https://github.com/iron-fish/ironfish/blob/master/ironfish/src/network/peers/peer.ts#L60-L64), and we are calling `getConnectedPeers`. This was originally just being used to convince the compiler of that, so really this should be a `getIdentityOrThrow`

Reply:
yeah that makes sense. ts compiler so weird sometimes.

Reply:
Ah actually thinking about this _more_, there is a possibility that the connection of the peer has changed since the sampling began, we should probably also verify that the peer is still in connected state at the top of the loop